EMP-240 ci using docker gitlab

raysiudak opened the pull request.

raysiudak committed:
Fix submodule repository url protocol to use shttp

The inotify submodule used ssh protocol to clone,
but this requires a public private key setup.

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add initial gitlab docker ci setup

This initial setup does a complete docker build
and makes sure the docker image is pushed.

It includes the gitlab ci setup file,
the dockerfile,
the check environment stub script,
the package build script.

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

raysiudak committed:
Add local build option and block pushing to the docker repo when not on a master branch.

dockerfile

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

oliv3r commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+*
+!tests/buildenv.sh
```
broken editor :p

oliv3r commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,105 @@
+lint package build script:
+  <<: *shell_linting
+  script:
+  - shellcheck -C -f tty -s sh build.sh
```
As this is very similar to my file :p I have this as well. We should think if we want to group these steps (e.g. by naming it 'lint shell scripts 1/2' and 2/2 the UI will group all the lint steps together. What do you think?

raysiudak replied:
Right now the naming says more about what it is doing, your suggestion hides that.

oliv3r replied:
sadly I agree, with only a small number of steps, not grouping it seems to be okay anyway. But we need to have a strategy for this in the future. Ex. if we start adding test, and we have 1 line for each test, we can have 100's of tests. Having them ungrouped makes the UI look bad imo. For example a combination of these two is done here: https://gitlab.com/gitlab-org/gitlab-ce/pipelines/25525772

raysiudak replied:
Yeah ok, but that is a problem when we actually have 100's of test (YAGNI?)

oliv3r commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,105 @@
+    name: "$CI_COMMIT_REF_NAME-$CI_COMMIT_SHA"
+    paths:
+    - ./*.deb
+    expire_in: 3 days
```
broken editor :p

oliv3r commented on .gitmodules:
```diff
@@ -1,3 +1,3 @@
 [submodule "inotify"]
 	path = inotify
-	url = git@github.com:dsoprea/PyInotify.git
+	url = https://github.com/dsoprea/PyInotify.git
```
note: we have to figure out how to do this with private modules: https://docs.gitlab.com/ee/ci/git_submodules.html suggests using relative urls, which is fine if it also works locally.

raysiudak replied:
hmm, this is a problem. You do not want to depend on a local relative git repo somewhere. Maybe we can make this conditional?

oliv3r replied:
lets discuss this at some point outside of this ticket, it was just a reminder for us :)
And I agree, I don't think relative URL's work for local cloning as it will interpret it as file://../..

oliv3r commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM registry.hub.docker.com/library/debian:stable-slim
+
+LABEL Maintainer="r.siudak@ultimaker.com" \
```
we may want to put something like software-platform@ultimaker.com here? what do you think?

raysiudak replied:
agree software-embedded-platform@ultimaker.com

oliv3r commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+RUN apt-get update && \
+    apt-get install -y \
+    python3-stdeb \
+    python-all
```
Can you convince me why this is not wrong?

See:
https://packages.debian.org/jessie/python3-all
https://packages.debian.org/jessie/python-all

Also, can you indent the packages with 4 more spaces and sort them? that way they are more obviously children of apt-get

raysiudak replied:
Nope if python3-all works than that is fine with me

oliv3r commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+      Comment="Ultimaker python3-inotify environment"
+
+RUN apt-get update && \
+    apt-get install -y \
```
Yesterday I think it was I realized we should also add no-install-recommends to not install recommended packages and arjen pointed out that -q makes the installation more quiet, ending up with apt-get install -q -y --no-install-recommends

oliv3r commented on README.md:
```diff
@@ -1,2 +1,6 @@
+[![pipeline status](https://gitlab.com/ultimaker_bv/um-python3-inotify/badges/master/pipeline.svg)](https://gitlab.com/ultimaker_bv/um-python3-inotify/commits/master)
```
note to ourselves; only merge this when we have our group up properly. As such, we should then thus move this package to ultimaker/embedded([-/]platform?)

oliv3r commented on build.sh:
```diff
@@ -0,0 +1,10 @@
+
+cd "${MODULE_PATH}"
+python3 setup.py --command-packages=stdeb.command bdist_deb
+cp deb_dist/python3-inotify_*.deb "${WORKDIR}/"
```
broken editor

oliv3r commented on build_for_ultimaker.sh:
```diff
@@ -1,20 +1,55 @@
 
-SRC_PATH="$(pwd)"
-MODULE_PATH="$(pwd)/inotify"
+CI_REGISTRY_IMAGE="${CI_REGISTRY_IMAGE:-registry.gitlab.com/ultimaker_bv/um-python3-inotify}"
```
note to ourselves, fix url

oliv3r commented on build_for_ultimaker.sh:
```diff
@@ -1,20 +1,55 @@
+fi
+
+echo "Starting build using ${CI_REGISTRY_IMAGE}:${CI_REGISTRY_IMAGE_TAG}."
+docker pull "${CI_REGISTRY_IMAGE}:${CI_REGISTRY_IMAGE_TAG}"
```
i just found out we can omit the pull as the run will do that anyway if it cannot find the requested image locally.

oliv3r commented on build_for_ultimaker.sh:
```diff
@@ -1,20 +1,55 @@
+
+echo "Starting build using ${CI_REGISTRY_IMAGE}:${CI_REGISTRY_IMAGE_TAG}."
+docker pull "${CI_REGISTRY_IMAGE}:${CI_REGISTRY_IMAGE_TAG}"
+fake_whoami
```
i'm happy with the fake whoami; but is anything in the package using it? it does not hurt to just have it there (and maybe we should just do this by default) what do you think?

raysiudak replied:
yeah maybe it is good too leave it in the template so that others do not run into the same issue if not having whoami in the docker container?

oliv3r replied:
:+1:

oliv3r commented on build_for_ultimaker.sh:
```diff
@@ -1,20 +1,55 @@
+echo "Starting build using ${CI_REGISTRY_IMAGE}:${CI_REGISTRY_IMAGE_TAG}."
+docker pull "${CI_REGISTRY_IMAGE}:${CI_REGISTRY_IMAGE_TAG}"
+fake_whoami
+docker run --rm -i -t -h "$(hostname)" -u "$(id -u)" \
```
i forgot to set the groupid myself as well, `-u "$(id -u):$(id -g)"`

oliv3r commented on build_for_ultimaker.sh:
```diff
@@ -1,20 +1,55 @@
+	   -v "${FAKE_WHOAMI}:/usr/bin/whoami" \
+	   -w "${WORKDIR}" \
+	   "${CI_REGISTRY_IMAGE}:${CI_REGISTRY_IMAGE_TAG}" \
+	   ./build.sh "${@}"
```
broken editor

oliv3r commented on tests/buildenv.sh:
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
```
is bash really required here? Try to avoid bashisms as long as possible to enhance portability.

The script doesn't do anything!

we should be able to check if we have a python interpreter and can import the stddeb stuff that we need from setup yeah?

probably easiest to do something like creating a python script and running it (or just copying a script if it is more then one line)

e.g.
```sh
#!/bin/sh

set -eux

TEST_FILE=$(mktemp --suffix=py)

cleanup() {
   if [ -e "${TEST_FILE}" ]; then
        rm "${TEST_FILE}"
}
trap cleanup EXIT

cat <<-EOT > ${TEST_FILE}
    import stdeb or whatever
    import the other thing

    maybe run something (init for example)
EOT

exit 0
```

echo "Checking python environment"
python3 "${TEST_FILE}"
(double check if the script returning an error aborts the script)

WesleyE commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+RUN apt-get update && \
+    apt-get install -y \
+    python3-stdeb \
+    python-all
```
You can get a smaller package by removing the package list afterward, this also increases the likelihood that the layer can be cached.

 && rm -r /var/lib/apt/lists/*

oliv3r replied:
also apt-get clean && \ (or rm -rf /var/cache/apt)

oliv3r commented on Dockerfile:
```diff
@@ -1,11 +1,12 @@
-    python-all
+    apt-get install -q -y --no-install-recommends \
+        python3-all \
+        python3-stdeb && \
```
to ease with copy/pasting, sorting and the sort, put the && \ on a new line. that way, you can do simply 'select packages, sort'

raysiudak replied:
Tried this, docker is unable to parse the line then because is does not recognize &&.

oliv3r commented on tests/buildenv.sh:
```diff
@@ -1,2 +1,10 @@
+
+python3 --version
+
+python3 -c "print('Dit is een {} test'.format('build env'))"
```
:rofl: i don't think this was intended :)

i like the use of -c; maybe do the two modules we build as imports before the test :)

oliv3r reviewed (changes requested):
A new think I learned from loy, is to use git commit --fixup <previous commitish you are addressing>

raysiudak committed:
fixup! Add initial gitlab docker ci setup
dockerfile

raysiudak committed:
Made some more improvements:

- Renamed gitlab build steps to make more sense
- Add minimal setup.py files to test the build env packaging
- Fixed issue that the buildenv script is run from the root directory,
  instead of the script root.
- Changed docker version to jessie

Contributes to: EMP-240

Signed-off-by: Raymond Siudak <r.siudak@ultimaker.com>

oliv3r reviewed (commented):
Surely you meant https :p

raysiudak committed:
fixup! fixup! Add initial gitlab docker ci setup dockerfile

oliv3r commented on tests/setup.py:
```diff
@@ -13,7 +13,7 @@ def __init__(self, *args, **kwargs):
     name='py3_only_pkg',
     packages=['py3_only_pkg'],
-    url='http://github.com/Ultimaker/python3-inotify/tests',
+    url=':https//github.com/Ultimaker/um-python3-inotify/tests',
```
what's the purpose of the : infront of the https?

oliv3r committed:
Reduce layers to improve efficiency

Having many COPY actions only adds in layers. We already decided to do
whitelisting in the dockerignore file, and thus do not need to do
further whitelisting in the COPY operation.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
Cleanup build env during start

Some build artifacts are left behind after each build, so start by
cleaning these up. This ensure stdeb can always run.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
Whitespace cleanup

Remove some stray tabs with spaces.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
Allow to work with only python3

While stdeb supports building python3 only packages, the build
requirement has an unneeded check for python-all. Patch this in our
current release until https://github.com/straw/stdeb/pull/140 is merged
and released.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

raysiudak committed:
fixup! fixup! fixup! Add initial gitlab docker ci setup dockerfile
